src/shadowbox/server/manager_service.spec.ts

```diff
@@ -30,7 +30,7 @@ describe('ShadowsocksManagerService', () => {
 
   it('lists access keys in order', (done) => {
     const repo = new MockAccessKeyRepository();
-    const service = new ShadowsocksManagerService(repo);
+    const service = new ShadowsocksManagerService(null, repo, null);
```

Comment:
Each call to the constructor in this test passes only a `repo`. DId you consider keeping that as the first argument and making the two new parameters optional?

Reply:
They are actually all mandatory. I added a new test case for renames that require the server config.

Reply:
More tests are good but I just meant that repeatedly constructing `ShadowsocksManagerService` with `(null, repo, null` suggests a better signature might be:
```ts
  constructor(
      private accessKeys: AccessKeyRepository,
      private serverConfig?: server_config.ServerConfig,
      private stats?: metrics.PersistentStats
  ) {}
```
Then your tests could create it with `ShadowsocksManagerService(repo)`.

Reply:
I kind of like to make it clear that we are ignoring some of the mandatory fields, so a reader is not misled to think everything is ok. It also makes it explicit that we are not exercising the code that use those parameters, which ideally we should.

Reply:
Hmm. I think you could justify almost any signature with that reasoning. Since 90% of the calls to the constructor come from tests - not unusual - I would ordinarily think it's worth spending a little time making it elegant.